python/hyperon/base.py

```diff
@@ -179,104 +237,198 @@ def get_atoms(self):
             result.append(Atom._from_catom(r))
         return result
 
-    """
-    Returns the Space object referenced by the SpaceRef, or None if the object does not have a
-    direct Python interface
-    """
     def get_payload(self):
+        """
+        Returns the Space object referenced by the SpaceRef, or None if the object does not have a
+        direct Python interface.
+        """
         return hp.space_get_payload(self.cspace)
 
-    """
-    Performs the specified query on the Space, and returns the result BindingsSet
-    """
     def query(self, pattern):
+        """
+        Performs the specified query on the Space, and returns the result as a BindingsSet.
+        """
         result = hp.space_query(self.cspace, pattern.catom)
         return BindingsSet(result)
 
-    """
-    Performs a substitution within the Space
-    """
     def subst(self, pattern, templ):
+        """
+        Performs a substitution within the Space, based on a pattern and a template.
+        """
         return [Atom._from_catom(catom) for catom in
                 hp.space_subst(self.cspace, pattern.catom,
                                          templ.catom)]
 
-"""
-A reference to a native GroundingSpace, implemented by the MeTTa core library
-"""
 class GroundingSpaceRef(SpaceRef):
+    """
+    A reference to a native GroundingSpace, implemented by the MeTTa core library.
+    This class extends SpaceRef to provide the same set of functionalities,
+    specifically for GroundingSpaces.
+    """
 
     def __init__(self, cspace = None):
+        """
+        Initialize a new GroundingSpaceRef.
+        If a CSpace object is provided, use it; otherwise create a new GroundingSpace.
```

Comment:
Do we *want* end users to be trying to interoperate between the python interface and the C interface?

It seems like they're asking for bugs and memory leaks if they try.  For example, if they take a C atom from a python Atom and use it to create an additional python Atom, and then they'll end up with a double-free.

My opinion is that the python API user should not know or care that C objects underly the python objects.  Python limits what we can do for data hiding, but I don't think we should document the interoperation between C objects and Python objects.

Reply:
> Do we want end users to be trying to interoperate between the python interface and the C interface?

I found this necessary for a few cases. While this is an indicator that some API functions are missing on the Python side, the user may want to address the C interface directly as a hotfix or to describe the issue, at least, at the moment.